### DISCUSSION TITLE
Reset vs setValue API

### DISCUSSION THREAD

**Discussion by mrlubos** [react-hook-form/react-hook-form#7671] [Q&A] [answered]
Hey everyone! I was wondering what's the conceptual difference between the `reset()` and `setValue()` APIs? Both require me to supply field name and value to reset to, correct? Why would I want to use one or the other method? Thanks!

**Comment by Moshyfawn** [comment, marked as answer]
The simplest rule of thumb is:
- `reset` is for **resting** the whole form or it's major part
- `setValue` is for **updating** a single field value

Looking further into semantics, `reset` will not only reset your form values, but it'll wipe the form's states like `isDirty`, or submission information like `isSubmitSuccessful` or `submitCount` **by default**. `setValue` allows for "manual" field value update, like updating some other field's value `onChange` of a different one without changing any form level state.

`reset` can set new defaults to the form's fields, `setValue` just updates a single field's value.

Say, you have remote data for an Edit form, that comes in after your form has been initialized. So, the form's gonna have some empty default values, that aren't the actual defaults for the data you wanna edit. In this case, when data comes, say, from BE, you'd `reset` the form's values as supposed to `setValue`, because you don't want your form to treat the BE data like a change, but rather as the starting values.

There're more intricacies to what `reset` and `setValue` do differently, but the info above might help you start seeing the differences.

**Reply by mrlubos** [reply]
Thanks @Moshyfawn! The thing I am thinking about is this: can I use `reset()` without specifying the value I am resetting to? If form knows the default value is `null` or whatever came from back-end, can I do something like `reset('field_name')` without knowing what the default value is?

**Reply by Moshyfawn** [reply]
I recommend you check out the official RHF docs on [the `reset` form method](https://react-hook-form.com/api/useform/reset) to learn about it's API.
You don't provide a field name to `reset`. It takes an object to reset the form to instead. If you don't provide any new default value to `reset`, the form's gonna be `reset` to the provided `defaultValues` at the `useForm` hook or `defaultValue` values on the component level, if provided.
```js
const { reset } = useForm({
  defaultValues: {
    name: "Default Name"
  }
})

useEffect(() => {
  // will result in form values reset to { name: "Default Name" } as provided at the useForm hook level via defaultValues
  reset()
}, [reset])
```

If you want to reset a single specific field, you want to look into using [the `resetField` form method](https://react-hook-form.com/api/useform/resetfield) instead. It takes the field's name to reset as it's first argument. It will reset the default value of the provided field, if not specified as an option otherwise. A bit similar to `setValue`, but you can specify the new default value of the field via `resetField` options.

You can take a look at [this CSB](https://codesandbox.io/s/dark-fire-d1xof?file=/src/App.tsx) for example. The defaultValues specified at the `useForm` hook level are gonna be the values the form resets to if you call `reset` without args. Calling `resetField` will do the same to the specified input if no new defaultValue provided via it's options. If you uncomment this part
```js
const {
  remoteDefaultValues,
  isRemoteDefaultValuesLoading
} = useRemoteDefaultValues();

useEffect(() => {
  if (!isRemoteDefaultValuesLoading && remoteDefaultValues) {
    reset(remoteDefaultValues);
  }
}, [reset, isRemoteDefaultValuesLoading, remoteDefaultValues]);
```
the `reset` method is gonna be called with the new "remote" default form values as it's argument after 2 seconds, which means, changing your form's input values and resetting them will result in field values set to the "remote" default values instead.

**Reply by mrlubos** [reply]
@Moshyfawn Perfect, thank you! `resetField()` is exactly what I needed. Thanks for clearing that up!

**Reply by bluebill1049** [reply]
Thank you @Moshyfawn 100% expert reply
